python/lsst/daf/butler/registry/dimensions/static.py

```diff
@@ -120,95 +121,196 @@ def initialize(
-        result = self.get(element)
-        assert result, "All records instances should be created in initialize()."
+    def fetch_cache_dict(self) -> dict[str, DimensionRecordSet]:
+        # Docstring inherited.
+        result: dict[str, DimensionRecordSet] = {}
+        with self._db.transaction():
+            for element in self.universe.elements:
+                if not element.is_cached:
+                    continue
+                assert not element.temporal, (
+                    "Cached dimension elements should not be spatial or temporal, as that "
+                    "suggests a large number of records."
+                )
+                if element.implied_union_target is not None:
+                    assert isinstance(element, Dimension), "Only dimensions can be implied dependencies."
+                    table = self._tables[element.implied_union_target.name]
+                    sql = sqlalchemy.select(
+                        table.columns[element.name].label(element.primary_key.name)
+                    ).distinct()
+                else:
+                    table = self._tables[element.name]
+                    sql = table.select()
+                with self._db.query(sql) as results:
+                    result[element.name] = DimensionRecordSet(
+                        element=element,
+                        records=[element.RecordClass(**row) for row in results.mappings()],
+                    )
         return result
 
+    def insert(
+        self,
+        element: DimensionElement,
+        *records: DimensionRecord,
+        cache: DimensionRecordCache,
+        replace: bool = False,
+        skip_existing: bool = False,
+    ) -> None:
+        # Docstring inherited.
+        if not element.has_own_table:
+            raise TypeError(f"Cannot insert {element.name} records.")
+        rows, overlap_insert_rows, overlap_delete_rows, overlap_summary_rows = self._make_record_db_rows(
+            element, records, replace=replace
+        )
+        table = self._tables[element.name]
+        with cache.modifying(element.name) as cache_records:
+            with self._db.transaction():
+                if replace:
+                    self._db.replace(table, *rows)
+                elif skip_existing:
+                    self._db.ensure(table, *rows, primary_key_only=True)
+                else:
+                    self._db.insert(table, *rows)
+                self._insert_overlaps(
+                    element, overlap_insert_rows, overlap_delete_rows, skip_existing=skip_existing
+                )
+                for related_element_name, summary_rows in overlap_summary_rows.items():
+                    self._db.ensure(self._overlap_tables[related_element_name][0], *summary_rows)
+            # Database transaction succeeded; update the cache to keep them
+            # consistent.
+            if cache_records is not None:
+                cache_records.update(records, replace=not skip_existing)
```

Comment:
Yeah I'm in favor of removing that complexity, would definitely simplify reasoning about this.  The refetch is still slightly tricky in theory because of cross-transaction visibility, but that's not a huge issue if the cache isn't shared between Butler instances.